hbase-backup/src/test/java/org/apache/hadoop/hbase/backup/master/TestBackupLogCleaner.java

```diff
@@ -61,8 +65,10 @@ public class TestBackupLogCleaner extends TestBackupBase {
 
   @Test
   public void testBackupLogCleaner() throws Exception {
```

Comment:
This test is not quite what you described in your earlier comment, and I'm pretty sure that it doesn't test the criteria in the same way either. It's a valid test case, but I think it's missing some cases that we would like to see covered.

In this test, we have a single region server, which serves as S2 from your above example.
1. At T0 you take a full backup B1 of all four tables at R1. `boundary` for our server is the timestamp of oldest of the wals of all the tables.
1. Data is inserted into tables 1 and 2.
1. At T1 you take incremental backup B1 of tables 1, 2, 3 to R1. `boundary` for our server doesn't change because the oldest wal is still from B1 and table 4. 
1. At T2 you take a full backup B2 of tables 1, 2, 3 to R2. `boundary` does not change because the oldest wal is still from B1 and table 4.

Based on my understanding, this test never sees a point when a wal can be discarded because wals from table4 are never rolled and so the wals identified for table4 in B1 are always the "oldest of the newest" for this server.

I think that we'd like to see a test that does what you described in your initial comment: observes the "rising tide" of wal timestamp on the server as backups across roots progress.

Reply:
I've majorly refactored & extended the testcase, and found a bug along the way.
@rmdmattingly @ndimiduk please take a look.